### PR TITLE
Check for shared before dataparallel call

### PR DESCRIPTION
**Pull request by hadasah**
**Patch description**
DataParallel objects get nested in at least these 5 files if __init__ is called with shared dict and data parallel is set to True. This isn't a problem is these nested DataParallel objects only observe, but it does become problematic if we want to call batch_act.

Fix is just to check for shared is None first

### CODE REVIEW

File: parlai/agents/bert_ranker/bi_encoder_ranker.py [hunk 1]
```diff
@@ -46,7 +46,7 @@ def __init__(self, opt, shared=None):
         super().__init__(opt, shared)
         # it's easier for now to use DataParallel when
         self.data_parallel = opt.get('data_parallel') and self.use_cuda
-        if self.data_parallel:
+        if self.data_parallel and shared is None:
```
Comment (emilydinan) [inline parlai/agents/bert_ranker/bi_encoder_ranker.py]: this is good but i'm just noticing that the `self.rank_loss` below is not correct. we should override `build_criterion` instead.

File: parlai/agents/bert_ranker/bi_encoder_ranker.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def __init__(self, opt, shared=None):
...
             self.model = torch.nn.DataParallel(self.model)
         if is_distributed():
```
Comment (emilydinan) [inline parlai/agents/bert_ranker/bi_encoder_ranker.py]: this should probably go under the check for `if self.data_parallel`..

File: parlai/core/torch_classifier_agent.py [hunk 3]
```diff
@@ -147,7 +147,8 @@ def __init__(self, opt, shared=None):
                     raise ValueError(
                         'Cannot combine --data-parallel and distributed mode'
                     )
-                self.model = torch.nn.DataParallel(self.model)
+                if shared is None:
```
Comment (emilydinan) [inline parlai/core/torch_classifier_agent.py]: let's get rid of this `if` here and put this entire `if self.use_cuda` block under the `else:` statement above

File: parlai/agents/bert_ranker/bi_encoder_ranker.py (hunk 2 above)
Comment (stephenroller) [reply parlai/agents/bert_ranker/bi_encoder_ranker.py]: ?

Comment (emilydinan) [reply parlai/agents/bert_ranker/bi_encoder_ranker.py]: the `if_distributed()` call raises an error and says distributed and data_parallel aren't compatible. but it doesn't make sense to raise this error `if not self.data_parallel`...

Comment (emilydinan) [reply parlai/agents/bert_ranker/bi_encoder_ranker.py]: (it is not related to the change that @hadasah made, it's an error i just noticed)

File: parlai/agents/bert_ranker/bi_encoder_ranker.py (hunk 1 above)
Comment (stephenroller) [reply parlai/agents/bert_ranker/bi_encoder_ranker.py]: Created a new issue: #2019.

File: parlai/agents/bert_ranker/bi_encoder_ranker.py (hunk 2 above)
Comment (stephenroller) [reply parlai/agents/bert_ranker/bi_encoder_ranker.py]: Created a new issue #2020.

File: parlai/core/torch_classifier_agent.py (hunk 3 above)
Comment (stephenroller) [reply parlai/core/torch_classifier_agent.py]: Let's actually do this one in this PR.

**Pull request merged**